src/twisted/web/http.py

```diff
@@ -1318,22 +1318,54 @@ def getAllHeaders(self):
         return headers
 
 
-    def getRequestHostname(self):
+    def getRequestHost(self):
         """
-        Get the hostname that the user passed in to the request.
+        Get the host that the client requested.
 
-        This will either use the Host: header (if it is available) or the
-        host we are listening on if the header is unavailable.
+        This will use the C{Host} header when it is available;
+        otherwise it will attempt to derive the host from the
+        transport's C{host}; transports that do not have host names,
+        like UNIX domain sockets, will raise L{UnsupportedTransport}.
 
         @returns: the requested hostname
+
         @rtype: C{bytes}
+
+        @raises: L{UnsupportedTransport} if there is no C{Host} header
+            and the transport does have a hostname (e.g., when
+            listening on a UNIX socket)
         """
-        # XXX This method probably has no unit tests.  I changed it a ton and
-        # nothing failed.
         host = self.getHeader(b'host')
         if host:
             return host.split(b':', 1)[0]
-        return networkString(self.getHost().host)
+        hostAddress = self.getHost()
+        if isinstance(hostAddress, (address.IPv4Address, address.IPv6Address)):
+            return networkString(self.getHost().host)
+        else:
+            raise UnsupportedTransport(
+                "Transport of type {} does not have a hostname.".format(
+                    type(hostAddress)))
+
+
+    def getRequestHostname(self):
+        """
+        Get the hostname that the user passed in to the request.
+
+        This will either use the Host: header (if it is available) or
+        the host we are listening on if the header is unavailable.
+
+        @returns: the requested hostname
+
+        @rtype: C{bytes}
```

Comment:
This does not match the return type the interface declares, which is `str`. I would prefer a native string here, as that is the type generally used for IP addresses and hostnames on Python 3.